src/data_structure/preprocess_data_structure.jl

```diff
@@ -177,6 +166,54 @@ function generate_direction()
     end
 end
 
+"""
+    process_loss_bidirectional_capacities()
+
+    For connections of type `:connection_type_lossless_bidirectional` if a `connection_capacity` is found
+    we ensure that it appies to each of the four flow variables
+
+"""
+
+function process_loss_bidirectional_capacities()
+    for c in connection(connection_type=:connection_type_lossless_bidirectional)
+        conn_capacity_param = nothing
+        found_from = false
+        for (n, d) in connection__from_node(connection=c)
+            if haskey(connection__from_node.parameter_values[(c, n, d)], :connection_capacity)
+                conn_capacity_param = connection__from_node.parameter_values[(c, n, d)][:connection_capacity]
+                found_from = true
```

Comment:
`haskey()` before accessing the dictionary means two dict lookups. Would you consider using `get` instead? (it only requires one lookup)

```julia
found_value = get(dict, key, nothing)
if found_value != nothing
    # use found_value directly
else
    # not in dictionary
end

```

Reply:
ah, ok - good to know

Reply:
I see that we look for connection_capacity only in parameter values, not in default values. Is that important?

We have the `indices` function that can collect keys where the value is different than `nothing`. It also accepts filters: 
```julia

for (c, n, d) in indices(connection_capacity; node=n, direction=d)
   ...
end
```

Maybe we can use it?

Also, I'm not a big fan of accessing and modifying the fields of `Parameter` values directly. SpineInterface should provide a function that does that. What do you think of something like `set_value!(parameter, class, index, value)`?

Reply:
Did you add any tests for this code? If yes, I think we can merge and I can do a bit of refactoring later...

Reply:
> I see that we look for connection_capacity only in parameter values, not in default values. Is that important?
> 
> We have the `indices` function that can collect keys where the value is different than `nothing`. It also accepts filters:
> 
> ```julia
> for (c, n, d) in indices(connection_capacity; node=n, direction=d)
>    ...
> end
> ```
> 
> Maybe we can use it?
> 
> Also, I'm not a big fan of accessing and modifying the fields of `Parameter` values directly. SpineInterface should provide a function that does that. What do you think of something like `set_value!(parameter, class, index, value)`?

There are two issues here... if I use indices(connection_capacity, ...) I don't know if what I'm getting back is on `connection__from_node` or `connection__to_node` - the way I have written it is the only way I could figure out to know on which relationships and which indices exactly I need to set connection_capacity.

The other issue is that I have no idea what the parameter may look like - when I go to the dicts directly I can just copy the parameter_value and it doesn't matter if it's a static value or a timeseries or whatever... so if we were to add support in SpineInterface for this sort of thing - we would also need `get_value_raw(parameter, class, index, value)` (or whatever)?

Reply:
> Did you add any tests for this code? If yes, I think we can merge and I can do this bit of refactoring later...

You mean the pre-processing code? Not yet - but I have implemented tests for all the new constraints. I did test that the new pre-processing code works correctly, and I will write some specific tests for it. I don't have time today - but perhaps we could merge anyway and I will do it on Monday? I always get a little anxious of slipping behind and having to deal with merge conflicts so it would be nice to get it in if we could?

Reply:
Let's merge and 'worry' about this later. You made fair points, and it looks like things need to get thought trough in SpineInterface. There's no point in delaying this PR because of that.

Everything else looks great, so I will finish my review so you can merge?